Use control comm target in base Manager

martinRenou opened the pull request:
Port of https://github.com/voila-dashboards/voila/pull/766 for JupyterLab and classic Jupyter Notebook

Use the `jupyter.widget.control` comm target for fetching widget states in the JupyterLab manager and the widgetsnbextension, this is used *e.g.* when refreshing the JupyterLab page on an already running kernel that contains widgets.

It will also be used by Voila in https://github.com/voila-dashboards/voila/pull/846.

I would really like this to be considered for the coming 8.0.0 release. It would also be nice to backport it to the 7.x branch if we keep making releases from there.

User facing changes
---------------------

This improves the performances when refreshing the JupyterLab page or when connecting to an already existing kernel that contains widgets.

martinRenou commented:
Making this WIP for now as I'll be doing the same for the widgetsnbextension

martinRenou commented:
@meeseeksbot backport to 7.x

martinRenou commented:
meeseeksdev backport to 7.x

martinRenou commented:
Comment: could this be part of the base manager?

martinRenou commented:
We could probably do this to check for the target:

```typescript
const reply = await kernel.requestCommInfo({target_name: this.comm_target_name});
```

The documentation is not entirely clear on the behavior of this when the target is no registered (to be improved):

https://jupyter-client.readthedocs.io/en/stable/messaging.html#comm-info

martinRenou commented:
Question: what happens if I reduce the tornado web socket message size to something very small? Does it fail? How?

vidartf commented on python/jupyterlab_widgets/src/manager.ts:
```diff
@@ -106,6 +121,106 @@ export abstract class LabWidgetManager
+      commId,
+      { widgets: null },
+      { version: CONTROL_COMM_VERSION }
+    );
```
Does this not need to be inside the `try` clause?

jasongrout commented:
> Question: what happens if I reduce the tornado web socket message size limit to something very small? Does it fail? How?

My guess is that the message will just never come back from the kernel. I think it would be good to have a timeout on the request that will fall back to the older method. Separately, I think it would be good to have a check in the server zmq/websocket bridge to see if a comm message is too big, and if it is, send a comm_close message both ways (to kernel and frontend) with an error status, indicating the comm channel had an error.

martinRenou commented on python/jupyterlab_widgets/src/manager.ts:
```diff
@@ -106,6 +121,106 @@ export abstract class LabWidgetManager
+    // Start creating all widgets
+    for (const [widget_id, state] of Object.entries(states) as any) {
+      try {
+        const comm = await this._create_comm('jupyter.widget', widget_id);
```
This can probably be faster with a `Promise.all` instead of a `for () {await [...];}`

martinRenou commented:
> We could probably do this to check for the target:
>
> ```typescript
> const reply = await kernel.requestCommInfo({target_name: this.comm_target_name});
> ```

Unfortunately this doesn't work. The kernel answers to the request with `{"status": "ok"}` even when there is no target handler registered.

martinRenou commented:
I updated the PR this morning:

- Moved the `loadFromKernel` logic into the base manager, as suggested by @vidartf 
- Fallback to the old implementation if the control target is not responding in time (4 seconds), as suggested by @jasongrout

Unfortunately we can't use the commInfo request to know if the target is registered kernel-side. See https://github.com/jupyter-widgets/ipywidgets/pull/3335#issuecomment-1005571385

martinRenou commented:
> Unfortunately this doesn't work. The kernel answers to the request with {"status": "ok"} even when there is no target handler registered.

Check again: what are the exact response messages when
- the comm target is not registered
- the comm target is registered

martinRenou commented:
So it's annoying, requesting the comm info will provide the same answer whether the target is registered or not:

```
{
    'comms': {},
    'status': 'ok'
}
```

No target registered:
--------------------

![no_target](https://user-images.githubusercontent.com/21197331/149168907-74763b5b-95e1-4f67-ae50-5a6e54060245.png)


Target registered:
-----------------

![target](https://user-images.githubusercontent.com/21197331/149168935-ec4e34d8-7e30-45a6-b474-317a187da496.png)


So using this request will not give us the information about the comm target.

martinRenou commented:
Also adding a link to the `comm_info_reply` implementation in ipykernel: https://github.com/ipython/ipykernel/blob/0ab288fe42f3155c7c7d9257c9be8cf093d175e0/ipykernel/kernelbase.py#L771-L787

martinRenou commented:
The xeus implementation looks similar https://github.com/jupyter-xeus/xeus/blob/eab12033ed8dc0202dfba733446dcf7ced10fca1/src/xkernel_core.cpp#L303-L322

jasongrout commented on packages/base-manager/src/manager-base.ts:
```diff
@@ -342,6 +361,197 @@ export abstract class ManagerBase implements IWidgetManager {
+   * If this fails (control comm handler not implemented kernel side),
+   * it will fallback to `_loadFromKernelSlow`.
+   */
+  protected async _loadFromKernel(): Promise<void> {
```
Can we add a note in the function documentation here that this is a utility method for subclasses to use as they see fit? Since we don't use this method in this base class, it is a bit confusing why it is defined here without extra context.

jasongrout replied:
(and same comment for the other loading function defined just below)

martinRenou replied:
Sure! Make sense. Thanks for the review!

jasongrout reviewed (commented):
Looks good! Two minor comments to address inline, and then +1 from me to merge. Thanks!

martinRenou commented on packages/base-manager/src/manager-base.ts:
```diff
@@ -342,6 +361,197 @@ export abstract class ManagerBase implements IWidgetManager {
+      const initComm = await this._create_comm(
+        CONTROL_COMM_TARGET,
+        uuid(),
+        { widgets: null },
```
This might be a leftover, I'll see if I can remove it

martinRenou replied:
Done :)

martinRenou commented:
Thanks for your review @jasongrout
